### ISSUE TITLE
Enable editing and sharing of user expressions

### ISSUE THREAD

**Issue by NyakudyaA** [qgis/QGIS#34690]
# Problem

The new save expressions in QGIS are very cool however it lacks a few options to make it awesome.

![user-expression](https://user-images.githubusercontent.com/2510900/75339700-b352a580-5899-11ea-9dd6-946765dd3612.png)

* Add ability to edit a user expression including the help text
* Add the ability to export user expression. 
* Import user expression - Expression differ between versions so a quick check would have to be done when importing. I assume add QGIS version in the XML for the user expression

*NB* On a side note maybe it's also worth to add the export/import options for Custom functions as well

**Comment by suricactus** [comment]
I would like to do this as I'm already playing with the expressions builder, but I have a few questions here:
### 1) Edit an expression
I think reusing the current code at `QgsExpressionStoreDialog` is better, than creating a new `QgsExpressionEditDialog`. To workaround the uniqueness of the name of the saved user expression, I'm planning to pass an empty `existingLabels` list so the check will always pass:
https://github.com/qgis/QGIS/blob/master/src/gui/qgsexpressionstoredialog.cpp#L36-L39

### 2) Import/Export expressions
What format is preferred to be used there? I saw the expressions are originally stored in `~/.local/share/QGIS/QGIS3/profiles/default/QGIS/QGIS3.ini`, but I find the format quite ugly for editing which might be the case when exported.
```ini
user\start_point%28geom%29\expression=start_point($geometry)
user\start_point%28geom%29\helpText="<!DOCTYPE HTML PUBLIC \"-//W3C//DTD HTML 4.0//EN\" \"http://www.w3.org/TR/REC-html40/strict.dtd\">\n<html><head><meta name=\"qrichtext\" content=\"1\" /><style type=\"text/css\">\np, li { white-space: pre-wrap; }\n</style></head><body style=\" font-family:'Noto Sans'; font-size:10pt; font-weight:400; font-style:normal;\">\n<p style=\" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;\">Starting point of a geometry</p></body></html>"
```
Is it ok if I reproduce the format already found at `./resources/function_help/json/*`? This will allow to add more fields in the future and in general is quite flexible? Something like:
```
{
    "qgis_version": "3.12.1",
    "created_at": "2020-01-01T01-01-01",
    "expressions": [
      {
        "name": "start_point(geom)",
        "type": "expression",
        "description": "Starting point of a geometry"
      }
    ]
}
```

**Comment by suricactus** [comment]
Duplicate of https://github.com/qgis/QGIS/issues/34690

**Comment by elpaso** [comment]
> @elpaso I saw you are the author of this feature. I would like to do this as I'm already playing with the expressions builder, but I have a few questions here:
> ### 1) Edit an expression
> 
> I think reusing the current code at `QgsExpressionStoreDialog` is better, than creating a new `QgsExpressionEditDialog`. To workaround the uniqueness of the name of the saved user expression, I'm planning to pass an empty `existingLabels` list so the check will always pass:
> https://github.com/qgis/QGIS/blob/master/src/gui/qgsexpressionstoredialog.cpp#L36-L39

You might better make that argument optional (const QStringList &existingLabels = QStringList() )


> ### 2) Import/Export expressions
> 
> What format is preferred to be used there?

JSON is the way to go.

> I saw the expressions are originally stored in `~/.local/share/QGIS/QGIS3/profiles/default/QGIS/QGIS3.ini`, but I find the format quite ugly for editing which might be the case when exported.
> 

> 
> Is it ok if I reproduce the format already found at `./resources/function_help/json/*`? This will allow to add more fields in the future and in general is quite flexible? Something like:
> 
> ```
> {
>     "qgis_version": "3.12.1",
>     "created_at": "2020-01-01T01-01-01",
>     "expressions": [
>       {
>         "name": "start_point(geom)",
>         "type": "expression",
>         "description": "Starting point of a geometry"
>       }
>     ]
> }
> ```

Yes, you can start from there, maybe add the username (as an "author" field)

Looking forward to your enhancements, they are very welcome!

**Comment by suricactus** [comment]
HI, @elpaso ,

Thanks for the feedback.

1) I am new in Cpp and I forgot that default arguments even exist. Your solution is much cleaner, indeed.
2) I have it working now for the import and export, you can check my branch https://github.com/suricactus/QGIS/commit/a7ad646e134c06a9e9d863d34ac213cff8b53ff3.

I still have to fix the documentation in the header file, but more or less this is the code. Here is a [video](https://imgur.com/26uclrC) how it works.

2.1) I'm planing to group the buttons for import and export under a single import/export dropdown button with menu entries for each operation (similar to how things are group together in the toolbar for the selection tool). 

2.2) Both of them will fire a file dialog where the json file is located.

2.3) Do you insist to keep the original storage implementation  in the `.ini` file? I think working with JSON is much easier and we can the `user_expressions.json` together with the other profile configuration files. Also this format mimics the one found in `resources/function_help/json/*convex_hull*`. Python functions are already stored outside the `~/.local/share/QGIS/QGIS3/profiles/default/QGIS/QGIS3.ini` file.

And this is the final structure of the JSON:
```
{
    "author": "Ivan Ivanov",
    "exported_at": "2020-03-10T14:41:34",
    "expressions": [
        {
            "description": "<!DOCTYPE HTML PUBLIC \"-//W3C//DTD HTML 4.0//EN\" \"http://www.w3.org/TR/REC-html40/strict.dtd\">\n<html><head><meta name=\"qrichtext\" content=\"1\" /><style type=\"text/css\">\np, li { white-space: pre-wrap; }\n</style></head><body style=\" font-family:'Noto Sans'; font-size:10pt; font-weight:400; font-style:normal;\">\n<p style=\" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;\">sums</p></body></html>",
            "expression": "1+1",
            "group": "user",
            "name": "sum",
            "type": "expression"
        },
        {
            "description": "<!DOCTYPE HTML PUBLIC \"-//W3C//DTD HTML 4.0//EN\" \"http://www.w3.org/TR/REC-html40/strict.dtd\">\n<html><head><meta name=\"qrichtext\" content=\"1\" /><style type=\"text/css\">\np, li { white-space: pre-wrap; }\n</style></head><body style=\" font-family:'Noto Sans'; font-size:10pt; font-weight:400; font-style:normal;\">\n<p style=\" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;\">sums 2</p></body></html>",
            "expression": "1+1 + 1",
            "group": "user",
            "name": "sum2",
            "type": "expression"
        }
    ],
    "qgis_version": "3.13.0-Master"
}
```

**Comment by havatv** [comment]
Expressions are supported in the just released version 0.15.0 of the* QGIS Resource Sharing* plugin if they area placed in the *expressions* folder of a collection.
Expressions are made available in *User expressions* in the expressions dialogue.
